weave/errors.py

```diff
@@ -1,11 +1,25 @@
+from typing import Optional, Iterable
+
 class WeaveUnmergableArtifactsError(Exception):
     pass
 
 
-class WeaveBaseError(Exception):
-    pass
+class WeaveFingerprintErrorMixin():
```

Comment:
you can do `_fingerprint = None` as the first line in the class and then you don't need to do the `hasattr` stuff

Reply:
That would be a class attribute which would make it accessible (and mutable) by all instances of that class. I could put this as an instance attribute in the __init__ method but then we'd have to call __init__ twice in the WeaveBaseError and make sure that any class inherits from the mixin calls __init__. I thought this would be easier and less code.